File: go/aws/transport_sqs.go
package aws

import (
	"context"
	"encoding/json"
	"errors"
	"sync"

	"github.com/aws/aws-lambda-go/events"
	"github.com/aws/aws-sdk-go-v2/aws"
	"github.com/aws/aws-sdk-go-v2/service/sqs"

	"github.com/sns45/anyhook/go/core"
)

// SqsTransport is core.Transport backed directly by AWS SQS (see doc.go for
// why this package calls aws-sdk-go-v2's SQS client directly rather than
// composing anyq's Go SQS producer/consumer).
//
// There is no consumer LOOP here: on AWS, delivery runs on Lambda's native
// SQS event-source mapping, which is already a push consumer. DispatchRecords
// is the seam NewDeliverHandler calls once per invocation, mapping each
// record 1:1 to one Engine.ProcessMessage call. A record whose handler
// returns an error is reported back via BatchItemFailures (Lambda's
// ReportBatchItemFailures partial-batch-failure mechanism) so ONLY that
// record is redelivered -- one record's failure must never fail the whole
// invocation, which under Lambda's default SQS semantics would redeliver
// every record in the batch and defeat per-message isolation (G3). Mirrors
// TS SqsTransport (transport-sqs.ts).
type SqsTransport struct {
	env Env

	mu      sync.Mutex
	handler func(ctx context.Context, m core.Message) error
}

// NewSqsTransport builds an SqsTransport over env.
func NewSqsTransport(env Env) *SqsTransport {
	return &SqsTransport{env: env}
}

// Send implements core.Transport: durably enqueues m onto the configured queue.
func (t *SqsTransport) Send(ctx context.Context, m core.Message) error {
	body, err := json.Marshal(m)
	if err != nil {
		return err
	}
	_, err = t.env.SQS.SendMessage(ctx, &sqs.SendMessageInput{
		QueueUrl:    aws.String(t.env.QueueURL),
		MessageBody: aws.String(string(body)),
	})
	return err
}

// Subscribe implements core.Transport: registers the engine's ProcessMessage
// handler, invoked per-record by DispatchRecords.
func (t *SqsTransport) Subscribe(_ context.Context, handler func(ctx context.Context, m core.Message) error) error {
	t.mu.Lock()
	t.handler = handler
	t.mu.Unlock()
	return nil
}

// DispatchRecords maps one SQS event batch (as delivered to a Lambda by the
// event-source mapping) to per-record engine deliveries, returning
// BatchItemFailures for any record whose handler returned an error or whose
// body failed to parse. handler overrides the Subscribe-registered handler
// when non-nil (used by tests and by handlers that want an explicit handler
// without going through Subscribe first).
func (t *SqsTransport) DispatchRecords(ctx context.Context, event events.SQSEvent, handler func(ctx context.Context, m core.Message) error) (events.SQSEventResponse, error) {
	h := handler
	if h == nil {
		t.mu.Lock()
		h = t.handler
		t.mu.Unlock()
	}
	if h == nil {
		return events.SQSEventResponse{}, errors.New("aws.SqsTransport.DispatchRecords called before Subscribe (and no handler was passed)")
	}

	var failures []events.SQSBatchItemFailure
	for _, record := range event.Records {
		var m core.Message
		if err := json.Unmarshal([]byte(record.Body), &m); err != nil {
			failures = append(failures, events.SQSBatchItemFailure{ItemIdentifier: record.MessageId})
			continue
		}
		if err := h(ctx, m); err != nil {
			failures = append(failures, events.SQSBatchItemFailure{ItemIdentifier: record.MessageId})
		}
	}
	return events.SQSEventResponse{BatchItemFailures: failures}, nil
}

var _ core.Transport = (*SqsTransport)(nil)

File: packages/aws/src/transport-sqs.ts
/** @fileoverview SqsTransport: composes @anyq/sqs's producer for send(); Lambda's SQS event source IS the push consumer (D2/P1, A5). @module @anyhook/aws */
import { createSQSProducer, type SQSProducer } from '@anyq/sqs';
import type { Transport, Message } from '@anyhook/core';
import type { SQSEvent, SQSBatchResponse } from 'aws-lambda';

export interface SqsTransportOptions {
  queueUrl: string;
  region?: string;
  endpoint?: string;
}

/**
 * anyhook `Transport` backed by AWS SQS. `send()` composes `@anyq/sqs`'s `createSQSProducer` (P1:
 * transport drivers are named/owned by anyq, never anyhook) for durable, at-least-once enqueue (G5).
 *
 * There is no anyq consumer LOOP here: on AWS, delivery runs on Lambda's native SQS event-source
 * mapping, which is already a push consumer anyq does not need to reimplement (A5). `dispatchRecords`
 * is the seam `deliverHandler` calls once per invocation, mapping each record 1:1 to one
 * `processMessage` call. A record whose handler throws is reported back via `batchItemFailures`
 * (Lambda's `ReportBatchItemFailures` partial-batch-failure mechanism) so ONLY that record is
 * redelivered — we deliberately do NOT let one record's throw fail the whole invocation, which under
 * Lambda's default SQS semantics would redeliver every record in the batch and defeat per-message
 * isolation (G3).
 */
export class SqsTransport implements Transport {
  private readonly producer: SQSProducer<Message>;
  private connected = false;
  private handler?: (m: Message) => Promise<void>;

  constructor(opts: SqsTransportOptions) {
    this.producer = createSQSProducer<Message>({
      queueUrl: opts.queueUrl,
      sqs: { region: opts.region ?? process.env.AWS_REGION ?? 'us-east-1', endpoint: opts.endpoint },
    });
  }

  private async ensureConnected(): Promise<void> {
    if (!this.connected) {
      await this.producer.connect();
      this.connected = true;
    }
  }

  async send(m: Message): Promise<void> {
    await this.ensureConnected();
    await this.producer.publish(m);
  }

  /** Registers the engine's `processMessage` handler; invoked per-record by `dispatchRecords`. */
  async subscribe(handler: (m: Message) => Promise<void>): Promise<void> {
    this.handler = handler;
  }

  /**
   * Map one SQS event batch (as delivered to a Lambda by the event-source mapping) to per-record
   * engine deliveries, returning `batchItemFailures` for any record whose handler threw.
   */
  async dispatchRecords(event: SQSEvent, handler?: (m: Message) => Promise<void>): Promise<SQSBatchResponse> {
    const h = handler ?? this.handler;
    if (!h) throw new Error('SqsTransport.dispatchRecords called before subscribe() (and no handler was passed)');

    const batchItemFailures: SQSBatchResponse['batchItemFailures'] = [];
    for (const record of event.Records) {
      try {
        await h(JSON.parse(record.body) as Message);
      } catch {
        batchItemFailures.push({ itemIdentifier: record.messageId });
      }
    }
    return { batchItemFailures };
  }
}
